shared/src/main/scala/mlscript/codegen/Symbol.scala

```diff
@@ -0,0 +1,168 @@
+package mlscript.codegen
+
+import mlscript.utils.shorthands._
+import mlscript.Type
+import mlscript.JSClassDecl
+import mlscript.MethodDef
+import mlscript.Term
+
+sealed trait LexicalSymbol {
+
+  /**
+    * The lexical name of the symbol. This is different from the runtime name,
+    * the name of the symbol in the generated code. We allow duplicates lexical
+    * names in the same scope.
+    */
+  def lexicalName: Str
+}
+
+sealed trait RuntimeSymbol extends LexicalSymbol {
+  def runtimeName: Str
+}
+
+sealed trait TypeSymbol extends LexicalSymbol {
+  val params: Ls[Str]
+  val actualType: Type
+}
+
+sealed class ValueSymbol(val lexicalName: Str, val runtimeName: Str) extends RuntimeSymbol {
+  override def toString: Str = s"value $lexicalName"
+}
+
+object ValueSymbol {
+  def apply(lexicalName: Str, runtimeName: Str): ValueSymbol =
+    new ValueSymbol(lexicalName, runtimeName)
+}
+
+sealed case class TypeAliasSymbol(
+    val lexicalName: Str,
+    val params: Ls[Str],
+    val actualType: Type
+) extends TypeSymbol
+    with LexicalSymbol {
+  override def toString: Str = s"type $lexicalName"
+}
+
+final case class BuiltinSymbol(val lexicalName: Str, feature: Str) extends RuntimeSymbol {
+  val runtimeName = lexicalName
+
+  override def toString: Str = s"function $lexicalName"
+
+  /**
+    * `true` if the built-in value had been accessed before.
+    * `Scope` will reuse the `lexicalName` if `accessed` is false.
+    */
+  var accessed: Bool = false
+}
+
+final case class StubValueSymbol(
+    override val lexicalName: Str,
+    override val runtimeName: Str,
+    /**
+      * Whether this stub is shadowing another symbol.
+      */
+    val shadowing: Bool,
+    previous: Opt[StubValueSymbol]
+)(implicit val accessible: Bool)
+    extends RuntimeSymbol {
+  override def toString: Str = s"value $lexicalName"
+}
+
+final case class ClassSymbol(
+    val lexicalName: Str,
+    val runtimeName: Str,
+    val params: Ls[Str],
+    val actualType: Type,
+    val fields: Ls[Str],
+    val methods: Ls[MethodDef[Left[Term, Type]]],
+) extends TypeSymbol
+    with RuntimeSymbol with Ordered[ClassSymbol] {
+
+  import scala.math.Ordered.orderingToOrdered
+
+  override def compare(that: ClassSymbol): Int = lexicalName.compare(that.lexicalName)
+
+  /**
+    * Fill up this field after the class translation.
+    */
+  var body: Opt[JSClassDecl] = N
+
+  /**
+    * Fill up this field after resolving classes.
+    */
+  var baseClass: Opt[ClassSymbol] = N
```

Comment:
Can you explain why `baseClass` and `body` need to be mutable variables? Why not have them as immutable parameters to the class?

Reply:
They were removed!